grudge/models/euler.py

```diff
@@ -0,0 +1,364 @@
+"""Grudge operators modeling compressible, inviscid flows (Euler)"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+
+from abc import ABCMeta, abstractmethod
+
+from dataclasses import dataclass
+from arraycontext import (
+    thaw,
+    dataclass_array_container,
+    with_container_arithmetic
+)
+
+from meshmode.dof_array import DOFArray
+
+from grudge.discretization import DiscretizationCollection
+from grudge.dof_desc import DOFDesc, DISCR_TAG_BASE, as_dofdesc
+from grudge.models import HyperbolicOperator
+from grudge.trace_pair import TracePair
+
+from pytools.obj_array import make_obj_array
+
+import grudge.op as op
+
+
+# {{{ Array container for the Euler model
+
+@with_container_arithmetic(bcast_obj_array=False,
+                           bcast_container_types=(DOFArray, np.ndarray),
+                           matmul=True,
+                           rel_comparison=True)
+@dataclass_array_container
+@dataclass(frozen=True)
+class EulerField:
+    mass: DOFArray
+    energy: DOFArray
+    momentum: np.ndarray
+
+    @property
+    def array_context(self):
+        return self.mass.array_context
+
+    @property
+    def dim(self):
+        return len(self.momentum)
+
+# }}}
+
+
+# {{{ Predefined initial conditions for the Euler model
+
+def vortex_initial_condition(x_vec, t=0):
+    """Initial condition adapted from Section 2 (equation 2) of:
+
+    - K. Mattsson, M. Sv\"{a}rd, M. Carpenter, and J. Nordstr\"{o}m (2006).
+    High-order accurate computations for unsteady aerodynamics.
+    [DOI](https://doi.org/10.1016/j.compfluid.2006.02.004).
```

Comment:
```suggestion
    `DOI <https://doi.org/10.1016/j.compfluid.2006.02.004>`__.
```
reST is weird, I know.

Reply:
Done! https://github.com/inducer/grudge/pull/196/commits/79eaac664a85a0d407a7e63c8042b6178f9d2d8e